### PR TITLE
Attribute Filters

### PR DESCRIPTION
**Pull request by stu-greenshoots**
This PR adds a new `FilterEngine` to the core codebase. Filters are small operations that can be performed on attributes and return a new attribute with an new value. 

The FilterEngine has 2 config files as with other engines so that core filters can be registered and, once published, custom filters can be defined.

Filters are separated from the attribute id (and accessors) in an attribute definition string using pipes (`|`) this means that filters can also be chained easily.

Filters can at the moment be used in message templates (eg `{user.first_name | uppercase}`) and in conditions (just added to the attribute definition in conditons)

Filters are referenced by their name - the core filters defined in the PR are:
- uppercase
- uppercase_first
- uppercase_words
- lowercase
- number_to_words
- ordinal
- ordinal_words

It is the intention that an attribute reference string can mention filters that do not exists or pipe attributes into filters of the wrong type and the flow will not break - only a warning level message is logged.

Reflection engine has been updated to account for the new engine, and the reflection API has been updated in the corresponding OD PR.

The ContextService has been updated to now accept filter arrays when getting attributes and attribute values - and a new method has been added for simplicity that takes a string attribute reference and handles the parsing itself

### CODE REVIEW

File: src/FilterEngine/Filters/Count.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    protected static ?string $componentDescription = 'Will return an int attribute with the count of the original attribute
+                                                      value.';
+
+    protected array $validAttributeTypes = [
+        StringAttributeValue::class
+    ];
+
+    protected function filter(Attribute $attribute): Attribute
+    {
+        return new IntAttribute($attribute->getId(), count($attribute->getValue()));
```
Comment (elliotmassen) [inline src/FilterEngine/Filters/Count.php:26]: How do we protect against `$attribute->getValue()` not being an array or `Countable`? If the user accidentally provides a string attribute for example, `count` will throw an error.

**Review by elliotmassen — COMMENTED** [review]
This is really great 🚀 I've pulled this locally and is all working as expected. Just a few thoughts below, but nothing major.

File: src/FilterEngine/Filters/Count.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+use OpenDialogAi\AttributeEngine\Attributes\IntAttribute;
+use OpenDialogAi\AttributeEngine\AttributeValues\StringAttributeValue;
+use OpenDialogAi\AttributeEngine\Contracts\Attribute;
+use OpenDialogAi\FilterEngine\Filters\BaseFilter;
+
+/**
+ *
+ */
+class Count extends BaseFilter
```
[the remark this answers is not in the document]
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/Count.php:13]: Yes, that's a good point, I'll update 👍

File: src/FilterEngine/Filters/BaseFilter.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use OpenDialogAi\Core\Components\ODComponentTypes;
+
+abstract class BaseFilter implements FilterInterface
+{
+    use ODComponent;
+
+    protected static string $componentType = ODComponentTypes::ATTRIBUTE_FILTER_TYPE;
+    protected static string $componentSource = ODComponentTypes::APP_COMPONENT_SOURCE;
+
+    protected array $validAttributeTypes = [];
```
[the remark this answers is not in the document]
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/BaseFilter.php:17]: No, I forgot to clean this up. I decided to err on the side of trying to work on any attribute type for now and catching any exceptions that arise.

I'll remove this for now, and we can think about stricter rules on filters when needed

File: src/FilterEngine/Filters/Count.php (hunk 1 above)
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/Count.php:26]: The flow may be a little hard to follow, but there is a method called `applyFilter` in the base class that wraps the call in a try catch and that is what is called by the service.

There is a testCount method in CodeFIltersTest that shows how a string attribute being run through count gets the same value back out

File: src/FilterEngine/Filters/BaseFilter.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    /**
+     * Applies the filter and returns a new attribute of the right type with the same name, but new value
+     *
+     * @param Attribute $attribute
+     * @return Attribute
+     */
+    abstract protected function filter(Attribute $attribute): Attribute;
```
[the remark this answers is not in the document]
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/BaseFilter.php:46]: I see your point, but at the moment, each filter returns a new attribute with the same ID as the original, so chaining is still valid as each step will return an attribute with the same name as the first. 

I would like to wait and see how this into other areas of the core codebase (expected attributes, actions etc) before making a refactor on this. 

I can see filters leaning more towards working just on attribute values going forward where the original values and names won't make a difference

File: src/FilterEngine/Filters/BaseFilter.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (elliotmassen) [reply src/FilterEngine/Filters/BaseFilter.php]: That makes sense - another thing that would be much easier with a switch to types by attribute values I think.

File: src/FilterEngine/Filters/Count.php (hunk 1 above)
Comment (elliotmassen) [reply src/FilterEngine/Filters/Count.php]: When testing via Webchat preview this doesn't seem to get caught:

Screenshot 2022-02-17 at 11 11 55

File: src/FilterEngine/Filters/BaseFilter.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (elliotmassen) [reply src/FilterEngine/Filters/BaseFilter.php:44]: I suppose I was thinking that if we introduce a wrapping input class now then it makes it easier to introduce other changes in the future without having to break the interface/signature. 

I agree that it will be easier once we switch to the emphasis of value types instead - however I still think there could be use of being able to store other kinds of data in the input. Like you say, at the moment this isn't something any of our filters required, but I'm thinking that it having it now might make it easier to introduce more complex filters down the line.

Another way of looking at it is that the current input constrains us to filters only acting on an attribute. So if in the future we want to add something like: `{user.shopping_list | join ", "}` we'd likely have to add some kind of `?array parameters = []` to the method signature here. If we already have a `FilterInput` then we can simply encapsulate this inside the input class.

It's likely a minor point - I'm more so taking inspiration from the issue we've had with attribute references becoming more complex, and the tendency toward using the `ParsedAttributeName` class rather than separately passing around context ID, attribute ID, accessors & filters. It's got me thinking that perhaps it's better to err on the side of wrapping inputs & outputs to keep signatures/interfaces simpler as things grow more complex.

File: src/FilterEngine/Filters/Count.php (hunk 1 above)
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/Count.php]: This was down to a difference between PHP versions. PHP 8 throws a type error where PHP 7 throws an InvalidArgumentException. Updaying the catch block to all Throwable fixes it

File: src/FilterEngine/Filters/BaseFilter.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (stu-greenshoots) [reply src/FilterEngine/Filters/BaseFilter.php]: This flow has been implemented now - all filters will take a FilterInput and return a FilterOutput. The service deals with taking the attribute from and output and passing in as part of a new input when chaining
